src/hotspot/share/classfile/classLoaderExt.cpp

```diff
@@ -284,14 +287,86 @@ InstanceKlass* ClassLoaderExt::load_class(Symbol* name, const char* path, TRAPS)
   }
   stream->set_verify(true);
 
-  ClassLoaderData* loader_data = ClassLoaderData::the_null_class_loader_data();
-  Handle protection_domain;
-  ClassLoadInfo cl_info(protection_domain);
-  InstanceKlass* k = KlassFactory::create_from_stream(stream,
-                                                      name,
-                                                      loader_data,
-                                                      cl_info,
-                                                      CHECK_NULL);
+  InstanceKlass* k = load_class_impl(name, path, CHECK_NULL);
+
+  SystemDictionaryShared::set_shared_class_misc_info(k, (ClassFileStream*)stream);
+
+  return k;
+}
+
+class URLClassLoaderTable : public ResourceHashtable<
+  Symbol*, Handle,
+  7, // prime number
+  ResourceObj::C_HEAP> {};
+
+static URLClassLoaderTable* _url_classloader_table = NULL;
+
+Handle ClassLoaderExt::create_url_classloader(Symbol* path, TRAPS) {
+  ResourceMark rm(THREAD);
+  JavaValue result(T_OBJECT);
+  Handle path_string = java_lang_String::create_from_str(path->as_C_string(), CHECK_NH);
+  JavaCalls::call_static(&result,
+                         vmClasses::jdk_internal_loader_ClassLoaders_klass(),
+                         vmSymbols::toFileURL_name(),
+                         vmSymbols::toFileURL_signature(),
+                         path_string, CHECK_NH);
+  assert(result.get_type() == T_OBJECT, "just checking");
+  oop url_h = result.get_oop();
+  objArrayHandle urls = oopFactory::new_objArray_handle(vmClasses::URL_klass(), 1, CHECK_NH);
+  urls->obj_at_put(0, url_h);
+
+  Handle url_classloader = JavaCalls::construct_new_instance(
+                             vmClasses::URLClassLoader_klass(),
+                             vmSymbols::url_array_classloader_void_signature(),
+                             urls, Handle(), CHECK_NH);
+  return url_classloader;
+}
+
+Handle ClassLoaderExt::create_and_add_url_classloader(Symbol* path, TRAPS) {
+  Handle url_classloader = create_url_classloader(path, CHECK_NH);
+  bool added = _url_classloader_table->put(path, url_classloader);
```

Comment:
added is not used. Did you want to assert that `added == true`?

I think it's better to move the body of this function to line 342. Then it would be clear that the entry doesn't exist (and there's no need to check for `added`).

Reply:
Modified per your suggestions and I don't need to keep the bool variable.